Improvement: 3093 Dashboard Services panel

alexis-ld opened the pull request:
**Component**: dashboard

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**:  #3093 

**Summary**: Adds a Services panel to the dashboard


https://user-images.githubusercontent.com/3480526/127501863-1663e0bd-936c-4812-8854-4f1b2bc2557b.mov



<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #3093 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

alexis-ld committed:
ui, translations: Add missing Translations for Services panel

Refs: #3093

alexis-ld committed:
shell-ui, alerts : Expose more alerts selectors

Refs: #3093

alexis-ld committed:
ui, dashboard: Add Services panel

Refs: #3093

alexis-ld committed:
ui, dashboard, tests: Remove unneeded imports in DashboardInventory tests

alexis-ld committed:
ui, tests: Add missing Router wrapper in custom render test method

This allows the use of react-router components insed tested components

Refs: #3093

alexis-ld committed:
ui, dashboard, tests: Add tests for Dashboard Services panel

Refs: #3093

bert-e commented:
# Hello alexis-ld,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @teddyandrieux

* @alexandre-allard-scality

* @jbwatenbergscality

* @lucieleonard

* @chengyanjin

* @gdemonet

* @thomasdanan

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,322 @@
+    }
+    return color;
+  }};
+`;
```
could be removed

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,322 @@
+                <label>Severity</label>
+                <NonHealthyPopUpSeverity status={status}>
+                  {status}
+                </NonHealthyPopUpSeverity>
```
Could benefit from the New `Text` component  in core-ui.  You may want to check the `statusText` in storybook

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,322 @@
+    return (
+      <NonHealthyServiceItemElement>
+        <Tooltip
+          // Right placement to avoid Z index issues between left sidebar and tooltip or out of screen tooltip displays
```
can be removed since we already fixed the issue ;)

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,322 @@
+  alerts,
+}: {
+  label: string,
+  status: string,
```
I believe the type of status can be more specific. 

We have already typed the Status in `AlertProvider`
> export type Status = 'healthy' | 'warning' | 'critical';

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,322 @@
+
+  const getStartDate = (alerts: Alert[]) => {
+    // Check we only have alerts and set startsAt properties
+    const filtered = alerts.filter((item) => item && item.startsAt);
+    // Sort to make sure we have oldest alert first
+    filtered.sort((a, b) => new Date(a.startsAt) - new Date(b.startsAt));
```
it's good thinking, however, I don't think it's necessary ;)
In theory, at the current time, the same alert should not be triggered more than once.

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,322 @@
+  const alertsLibrary = useAlertLibrary();
+
+  // K8s Master
+  const k8sAlerts = useHighestSeverityAlerts(
```
In this specific case, we only gather two alerts: "K8sMasterServicesAtRisk" and "K8sMasterServicesDegraded"
and then we filter out the lower severity, so in the end, we should have at most one alert or an empty array. 
```suggestion
  const k8sHighestSeverityAlert = useHighestSeverityAlerts(
```

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,322 @@
+}: {
+  label: string,
+  status: string,
+  alerts: Alert[],
```
IMO, we should have at most one alert, please see the explanation below :)
suggestion:
`alert: Alert,`

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,322 @@
+          }}
+          overlay={
+            <NonHealthyPopUp>
+              <NonHealthyPopUpTitle>View details</NonHealthyPopUpTitle>
```
I think we should translate this as well.

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,322 @@
+            <NonHealthyPopUp>
+              <NonHealthyPopUpTitle>View details</NonHealthyPopUpTitle>
+              <NonHealthyPopUpItem>
+                <label>Severity</label>
```
missing translation.

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,322 @@
+                // Checking date validity
+                date instanceof Date && !isNaN(date) && (
+                  <NonHealthyPopUpItem>
+                    <label>Start</label>
```
ideally, we should translate start as well.

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,322 @@
+                  <NonHealthyPopUpItem>
+                    <label>Start</label>
+                    <div>
+                      {`${date.toISOString().split('T')[0]}
+              ${date.toISOString().split('T')[1].slice(0, 8)}`}
```
I think second is not necessary.
so you could benefit from`formatDateToMid1` in utils.js

ChengYanJin commented on ui/src/components/__TEST__/util.js:
```diff
@@ -58,15 +62,17 @@ const AllTheProviders = ({ children }) => {
-        </QueryClientProvider>
-      </Provider>
-    </IntlProvider>
+    <Router history={history}>
```
can you please explain to me why we need this Wrapper?

alexis-ld replied:
@ChengYanJin , because we use react-router-dom's `Link` in the component, without the Router wrapper this makes the test suite fail because react-router component cannot be rendered outside a Router context.

Here we create an in-memory fake history to be provided to the test suite.

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,322 @@
+  useHighestSeverityAlerts,
+  highestAlertToStatus,
+} from '../containers/AlertProvider';
+import CircleStatus from './CircleStatus';
```
I thought we already move it core-ui, but apparently not.. ;) 
Not the scope for this PR will do it in another PR.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @teddyandrieux

* @alexandre-allard-scality

* @jbwatenbergscality

* @lucieleonard

* @chengyanjin

* @gdemonet

* @thomasdanan





The following reviewers are expecting changes from the author, or must review again:

* @chengyanjin

alexis-ld committed:
ui, dashboard: Refactor Dashboard service to leverage existing utils and core-ui's components

Leverage core-ui's StatusText instead of local implementation
Use formatDateToMid1 instead of custom Date formatting

Refs: #3093

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,295 @@
+              </NonHealthyPopUpItem>
+              {
+                // Checking date validity
+                date && (
```
```suggestion
                alert[0].startsAt && (
```
if we want to make sure we have a startsAt property, we can check here.

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,295 @@
+  const k8sStatus = highestAlertToStatus(k8sHighestSeverityAlert);
+
+  // Bootstrap
+  const bootstrapAlerts = useHighestSeverityAlerts(
```
```suggestion
  const bootstrapHighestSeverityAlert = useHighestSeverityAlerts(
```
same for the rest...

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,295 @@
+
+  const getStartDate = (alerts: Alert[]) => {
+    // Check we only have alerts and set startsAt properties
+    const filtered = alerts.filter((item) => item && item.startsAt);
```
I think the filter is not necessary here, since what you want is only a check to make sure we do have the startsAt property.

ChengYanJin commented on ui/src/components/DashboardServices.js:
```diff
@@ -0,0 +1,295 @@
+        </ServiceItemLabelWrapper>
+      </ServiceItemElement>
+    );
+  else
```
if you want to be super strict,  but I think it's not necessary.. 
```suggestion
  else if(alerts.length === 1  && status !== STATUS_HEALTH)
```
and in the end, if we are not in any of the two scenarios, 
else{
 console.error(` The Service ${label} status has encountered an expected scenario.`)
}

ChengYanJin reviewed (approved):
LGTM!

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nicolast

* @teddyandrieux

* @alexandre-allard-scality

* @jbwatenbergscality

* @lucieleonard

* @chengyanjin

* @gdemonet

* @thomasdanan

ChengYanJin commented:
/bypass_author_approval

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.11`



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **bypass_author_approval**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.11`



The following branches have **NOT** changed:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue None.

Goodbye alexis-ld.
